demo.rhm

```diff
@@ -478,8 +480,8 @@ bob_loc
 def Map(symbol(alice), alice_loc2, symbol(bob), bob_loc2): local_map
 [alice_loc2, bob_loc2]
 
-def [ps :: Posn, ...] : [Posn(1, 2), Posn(3, 4)]
-ps[0].x
+def [p :: Posn, ...] : [Posn(1, 2), Posn(3, 4)]
+[p, ...][0].x
```

Comment:
For subscripting with `[0]`, it would make sense to me to let `p[0]` refer to the first element of the repetition `p`.

Can array access accept a repetition without creating ambiguity?

Reply:
Perhaps the repetitions inside array access could be restricted to just variables and further nested array access.

Reply:
This case shows a conflict:
```
def [ps :: List.of(Posn), ...] : [[Posn(1, 2), Posn(3, 4)], [Posn(5, 6), Posn(7, 8)]]
[length(ps[0]) + ps[0].x, ...]
```
If it's greedy from the bottom up, then the `length(ps[0])` will be fine as a single access but the `ps[0].x` will be wrong.
If it goes from the top down, then the `length(ps[0])` will be wrong while the `ps[0].x` is a repeated access.

Reply:
Allowing a repetition before `[0]` maybe could probably work, but with the current implementation, it would have to be limited somehow.

The intent with the current design is that a term is in an expression position or a repetition position based on its context. The implementation relies on that constraint; concretely, an attempt to match the `:repetition` syntax class internally either succeeds with an expansion or fails with an error, and the same is true of `:expression`. (In contrast, `:definition` is set up to fail with a non-match if the immediate term doesn't start a definition form, so it makes sense to try `:definition` and then `:expression` in a body sequence.) As you suggest, there would have to be some sort of constraints on either repetitions or expressions in an overlapping position so that expansion could pick the right one.

More generally, though, I think your example of ambiguity points to the difficulty of making two different kinds of inference work together: inferring a repetition versus an expression and inferring a repetition from an expression-like term in a repetition position. I'm not really sure either of those is a good idea, although the latter seems more useful, and I would be inclined (so far) to avoid the former.